QueryString and SimpleQueryString Graph Support

mattweber opened the pull request:
Add support for graph token streams to "query_string" and
"simple_query_string" queries.

mattweber commented:
There is a easily fixable conflict on cherry-picking this into `5.x`.  I have a branch ready for a PR against that if needed.

mattweber commented:
Removed the note in `synonym_graph` docs saying that only `match` queries are supported.  I believe all analyzed queries are supported once this PR makes it in.

@mikemccand @jimczi  can you please review?

mikemccand commented:
Thanks @mattweber I will review!

jimczi commented on core/src/main/java/org/elasticsearch/common/lucene/search/Queries.java:
```diff
@@ -135,6 +137,41 @@ public static Query applyMinimumShouldMatch(BooleanQuery query, @Nullable String
+            // we have a graph query that has at least one boolean sub-query
+            // re-build and set minimum should match value on all boolean queries
+            List<Query> oldQueries = ((GraphQuery) query).getQueries();
+            Query[] queries = new Query[oldQueries.size()];
```
Maybe just:
````
 queries[i] = maybeApplyMinimumShouldMatch(oldQueries.get(i));
````
... this would avoid the duplicate code below ?

jimczi commented on core/src/main/java/org/elasticsearch/common/lucene/search/Queries.java:
```diff
@@ -135,6 +137,41 @@ public static Query applyMinimumShouldMatch(BooleanQuery query, @Nullable String
+            List<Query> oldQueries = ((GraphQuery) query).getQueries();
+            Query[] queries = new Query[oldQueries.size()];
+            for (int i = 0; i < queries.length; i++) {
+                Query oldQuery = oldQueries.get(i);
```
Maybe just:
````
queries[i] = maybeApplyMinimumShouldMatch(oldQueries.get(i));
`````
... this would avoid the duplicate code below ?

mattweber commented:
Thanks @jimczi Let me look into the multi-match query and get that fix in this PR if needed.

mattweber commented:
@jimczi addressed your comments, can you please have another look?  Good catch on the `MultiMatchQuery#parseAndApply`, I had totally missed that in earlier PR's.

jimczi commented on core/src/test/java/org/elasticsearch/search/query/QueryStringIT.java:
```diff
@@ -263,6 +257,82 @@ public void testAllFieldsWithSpecifiedLeniency() throws IOException {
+        assertHitCount(searchResponse, 3L);
+        assertSearchHits(searchResponse, "1", "2", "6");
+    }
+
```
Can you add a test with `splitOnWhitespace(true)`  as well ? Something that checks that single term synonyms that rewrite to multi terms are handled correctly.

mattweber commented:
Thanks @jimczi I will add that test.

mattweber commented:
@jimczi test added.  Thanks for reviewing!

mikemccand commented:
LGTM

Thanks @mattweber and @jimczi!

mattweber commented:
Thanks @mikemccand.  I have squashed commits so it is ready for merge.  I will work on a `5.x` PR with the changes from earlier.

jimczi commented:
Thanks @mattweber @mikemccand  ! This is now pushed to master and 5.x.
